File: app/models/survey.py
from typing import Optional, Any
from pydantic import BaseModel, Field

class ColumnMapping(BaseModel):
    point_id: str = Field(..., description="Column name or index for Point ID")
    x: str = Field(..., description="Column name or index for X / Easting")
    y: str = Field(..., description="Column name or index for Y / Northing")
    rl: str = Field(..., description="Column name or index for Reduced Level / Elevation / Z")
    code: Optional[str] = Field(None, description="Optional column name or index for Code / Description")

class RawPreviewResponse(BaseModel):
    file_id: str
    filename: str
    file_type: str
    file_size_bytes: int
    total_rows: int
    total_columns: int
    has_headers: bool
    headers: list[str]
    detected_columns: ColumnMapping
    column_confidence: dict[str, float]
    preview_rows: list[dict[str, Any]]
    suggested_crs: Optional[str] = None

class ValidationIssue(BaseModel):
    row_index: int
    point_id: Optional[str] = None
    issue_type: str  # "missing_rl", "null_xy", "non_numeric", "duplicate_xy_conflict", "duplicate_id", "outlier"
    message: str
    raw_values: dict[str, Any]

class SurveyBounds(BaseModel):
    min_x: float
    max_x: float
    min_y: float
    max_y: float
    min_z: float
    max_z: float
    range_x: float
    range_y: float
    range_z: float

class CRSDetails(BaseModel):
    epsg: Optional[str] = None
    name: Optional[str] = None
    is_projected: Optional[bool] = None
    unit: str = "meter"
    datum: Optional[str] = None
    status: str  # "CONFIRMED", "UNCONFIRMED", "LOCAL_GRID"
    warning: Optional[str] = None

class ValidationSummary(BaseModel):
    total_records: int
    valid_points: int
    invalid_records: int
    duplicate_xy_count: int
    missing_rl_count: int
    duplicate_id_count: int
    bounds: Optional[SurveyBounds] = None
    crs: CRSDetails
    issues: list[ValidationIssue]
    preview_valid_rows: list[dict[str, Any]]
    preview_invalid_rows: list[dict[str, Any]]

class ValidateRequest(BaseModel):
    file_id: str
    column_mapping: ColumnMapping
    has_headers: bool = True
    source_crs: Optional[str] = None
    vertical_datum: Optional[str] = "Local TBM"
    horizontal_unit: str = "meter"
    vertical_unit: str = "meter"
    accept_valid_only: bool = False

class ProjectMetadata(BaseModel):
    project_id: str
    project_name: str
    created_at: str
    updated_at: str
    source_file: str
    file_size_bytes: int
    source_crs: Optional[str]
    processing_crs: Optional[str]
    elevation_datum: str
    horizontal_unit: str
    vertical_unit: str
    column_mapping: ColumnMapping
    point_count: int
    valid_point_count: int
    rejected_point_count: int
    bounds: Optional[SurveyBounds] = None
    software_version: str = "1.0.0"

File: frontend/src/types/survey.ts
export interface ColumnMapping {
  point_id: string;
  x: string;
  y: string;
  rl: string;
  code?: string | null;
}

export interface RawPreviewResponse {
  file_id: string;
  filename: string;
  file_type: string;
  file_size_bytes: number;
  total_rows: number;
  total_columns: number;
  has_headers: boolean;
  headers: string[];
  detected_columns: ColumnMapping;
  column_confidence: Record<string, number>;
  preview_rows: Record<string, any>[];
  suggested_crs?: string | null;
}

export interface ValidationIssue {
  row_index: number;
  point_id?: string | null;
  issue_type: string;
  message: string;
  raw_values: Record<string, any>;
}

export interface SurveyBounds {
  min_x: number;
  max_x: number;
  min_y: number;
  max_y: number;
  min_z: number;
  max_z: number;
  range_x: number;
  range_y: number;
  range_z: number;
}

export interface CRSDetails {
  epsg?: string | null;
  name?: string | null;
  is_projected?: boolean | null;
  unit: string;
  datum?: string | null;
  vertical_datum?: string | null;
  status: 'CONFIRMED' | 'UNCONFIRMED' | 'LOCAL_GRID';
  warning?: string | null;
}

export interface ValidationSummary {
  total_records: number;
  valid_points: number;
  invalid_records: number;
  duplicate_xy_count: number;
  missing_rl_count: number;
  duplicate_id_count: number;
  bounds?: SurveyBounds | null;
  crs: CRSDetails;
  vertical_datum?: string | null;
  issues: ValidationIssue[];
  preview_valid_rows: Record<string, any>[];
  preview_invalid_rows: Record<string, any>[];
}


export interface CRSPreset {
  code: string;
  name: string;
  type: string;
  unit: string;
}

export interface ValidateRequest {
  file_id: string;
  column_mapping: ColumnMapping;
  has_headers: boolean;
  source_crs?: string | null;
  vertical_datum?: string | null;
  horizontal_unit?: string;
  vertical_unit?: string;
  accept_valid_only?: boolean;
}
